### ISSUE TITLE
train

### ISSUE THREAD

**Issue by jl749** [jl749/YOLOv3#3]
## train.csv contains 16551 samples | test.csv contains 4952 samples
BATCH_SIZE = 32
16551/32 ≓ 518 loops per epoch
4952/32 ≓ 155 loops per epoch

**Comment by jl749** [comment]
am I using CUDA??  -- check --> `watch -n 0.5 nvidia-smi`
![image](https://user-images.githubusercontent.com/67103130/150742743-d80cf6c5-6224-422b-a131-53252a0280ab.png)

**Comment by jl749** [comment]
## Automatic Mixed Precision
float16 forward pass, accuracy speed trade off

![image](https://github.com/jl749/YOLOv3/assets/67103130/eb6a1975-99e6-4e82-bcae-a28d4d75d5bd)
1. create copy of fp32 weight
2. use copied weight to perform forward pass
3. cast fp16 prediction tensor back to fp32 
4. calculate fp32 loss, multiply **scaling factor** S 
5. scaled fp32 loss --> fp16 (prevent 0 clipping issue as shown in no.6)
6. ![image](https://github.com/jl749/YOLOv3/assets/67103130/24a9ae3f-4bde-478f-8b6a-cadfdf5223cc)
7. use scaled fp16 to do backward propagation and calculate gradients
8. fp16 gradients --> fp32 gradients / **scaling factor** S
9. update fp32 weight tensor using fp32 gradient



**REFERENCE**
https://pytorch.org/docs/stable/amp.html
https://pytorch.org/docs/stable/notes/amp_examples.html
https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html
https://openreview.net/pdf?id=r1gs9JgRZ
https://www.youtube.com/watch?v=b5dAmcBKxHg
